components/core/src/compressor_frontend/InputBuffer.hpp

```diff
@@ -0,0 +1,82 @@
+#ifndef COMPRESSOR_FRONTEND_INPUT_BUFFER_HPP
+#define COMPRESSOR_FRONTEND_INPUT_BUFFER_HPP
+
+// Project Headers
+#include "../ReaderInterface.hpp"
+#include "Buffer.hpp"
+
+namespace compressor_frontend {
+    class InputBuffer : public Buffer<char> {
+    public:
+
+        /**
+         * Resets input buffer
+         * @return
+         */
+        void reset () override;
+
+        /**
+         * Checks if reading into the input buffer won't overwrite data not yet used
+         * (e.g., data being overwritten is already compressed in the case of compression)
+         * @return bool
+         */
+        bool read_is_safe ();
+
+        /**
+         * Reads into the half of the buffer currently available
+         * @param reader
+         */
+        void read (ReaderInterface& reader);
+
+        /**
+         * Reads if no unused data will be overwritten
+         * @param reader
+         */
+        void try_read (ReaderInterface& reader) {
+            if (read_is_safe()) {
+                read(reader);
+            }
+        }
+
+        /**
+         * Swaps to a dynamic buffer (or doubles its size) if needed
+         * @return bool
+         */
+        bool increase_size_and_read (ReaderInterface& reader, size_t& old_storage_size);
```

Comment:
* Seeing as the size used everywhere else is `uint32_t`, why use `size_t` here?
* Typically, the capacity of a buffer is called its capacity rather than its size. So how about `increase_capacity_and_read`?